### PR TITLE
allow negative offset

### PR DESCRIPTION
**Pull request by pschou**
See issue #8458 for comments

### CODE REVIEW

**Comment by roidelapluie** [comment]
It seems that your branch is out of sync with master and has conflicts.

Additionally, I'd suggest you start of the following branch/commit and add the necesarry feature flag that was missing into that pull request: https://github.com/prometheus/prometheus/pull/6283

Feature flags are managed in main.go (look for `remove-write-receiver` there.)

Thanks!

**Comment by roidelapluie** [comment]
cc @dgl for awareness

**Comment by pschou** [comment]
Ok, rebasing here...

**Comment by pschou** [comment]
@roidelapluie Please let me know how this looks for a negative offset.

**Comment by roidelapluie** [comment]
Can we moved the unrelated docs change in a separate pull request? Thanks!

**Comment by pschou** [comment]
Yes, we can.

All the best!

**Comment by roidelapluie** [comment]
promql/parser/generated_parser.y.go is conflicting.

Can you please add a feature flag?

**Comment by pschou** [comment]
Please note that the file generated_parser.y.go is the output of `goyacc`, so it isn't a breaking change.  The source file is generated_parser.y.  

I can remove generated_parser.y.go if you would like as it will be compiled again by the deployment process.  I provide the compiled file only as a courtesy as it was pre-compiled when I cloned the repo.  :)

**Comment by codesome** [comment]
> I can remove generated_parser.y.go if you would like as it will be compiled again by the deployment process.

We actually need the file for the builds and tests to pass. Not sure why the conflict, can you try running goyacc again and push?

**Comment by roidelapluie** [comment]
Can you please rebase your pull request on top of master and regenerate it?

Le mar. 16 févr. 2021 à 07:37, Ganesh Vernekar <notifications@github.com> a
écrit :

File: promql/parser/generated_parser.y [hunk 1]
```diff
@@ -376,7 +376,12 @@ paren_expr      : LEFT_PAREN expr RIGHT_PAREN
 
 offset_expr: expr OFFSET duration
                         {
-                        yylex.(*parser).addOffset($1, $3)
+                        yylex.(*parser).addOffset($1, $3, 1)
+                        $$ = $1
+                        }
+                | expr OFFSET SUB duration
```
Comment (pschou) [inline promql/parser/generated_parser.y:382]: I can rebase and rebuild, but there will always be differences in the goyacc output due to these additional five lines, mainly adding this or | operator and this section to handle the 'OFFSET SUB duration' in opposed to just 'OFFSET duration' section.

**Review by pschou — COMMENTED** [review]
The output of goyacc will always be significantly different due to it being compiled.

For comparison: in #8121 there is a large blob commit for the same file for just this reason promql/parser/generated_parser.y.go

File: promql/parser/generated_parser.y (hunk 1 above)
Comment (codesome) [reply promql/parser/generated_parser.y:382]: Yes the difference expected, the conflict is mostly due to the commit history having more changes in the master that you dont have in your branch.

**Review by codesome — COMMENTED** [review]
We want this to be behind a feature flag, `--enable-feature=negative-offset`. Can you make those changes, please? You can take a hint from https://github.com/prometheus/prometheus/pull/8121 on how to pass and validate this option (main.go and engine.go, `validateOpts()` in engine.go).

Accordingly, you will need to update the docs in disabled_features.md for this new option and basics.md mentioning that it is disabled and how to enable.

Also, can you fix the DCO? Thanks!

File: promql/engine.go [hunk 2]
```diff
@@ -388,33 +393,49 @@ func (ng *Engine) newQuery(q storage.Queryable, expr parser.Expr, start, end tim
...
+	if ng.enableAtModifier && ng.enableNegativeOffset {
 		return nil
 	}
 
 	var validationErr error
 	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
 		switch n := node.(type) {
 		case *parser.VectorSelector:
-			if n.Timestamp != nil || n.StartOrEnd == parser.START || n.StartOrEnd == parser.END {
+			if !ng.enableAtModifier &&
```
Comment (roidelapluie) [inline promql/engine.go:407]: This is becoming a bit unreadable.
Can we go for a pattern:
```
var atModifierUsed, negativeOffsetUsed
```

and at the end

```
if  atModifierUsed && !ng.enableAtModified {
 return err
}
```

Comment (pschou) [reply promql/engine.go]: Yes, done.

File: promql/parser/generated_parser.y [hunk 3]
```diff
@@ -383,7 +383,12 @@ paren_expr      : LEFT_PAREN expr RIGHT_PAREN
...
+                        {
+                        yylex.(*parser).addOffset($1, $4, -1)
```
Comment (roidelapluie) [inline promql/parser/generated_parser.y:391]: Can we do this instead?

```suggestion
                        yylex.(*parser).addOffset($1, -1*$4)
```

**Comment by roidelapluie** [comment]
Can you please rebase (docs/disabled_features.md) is conflicting?

Also, can you please update promql/parser/printer_test.go with a test ? It seems not to work out of the box.

**Comment by pschou** [comment]
> Can you please rebase (docs/disabled_features.md) is conflicting?
> 
> Also, can you please update promql/parser/printer_test.go with a test ? It seems not to work out of the box.

yup, done

File: docs/disabled_features.md [hunk 4]
```diff
@@ -5,18 +5,33 @@ sort_rank: 10
 
 # Disabled Features
 
-Here is a list of features that are disabled by default since they are breaking changes or are considered experimental.
```
Comment (roidelapluie) [inline docs/disabled_features.md:8]: Can you please revert these changes? All the pages in the docs use "you". We should not change the pattern for just une file and changing all the doc is out of scope for this PR.

Comment (pschou) [reply docs/disabled_features.md]: No problem!  Reverted.

**Comment by roidelapluie** [comment]
I will look at this later, can you please fix the DCO?

cc @codesome do you want to review this as well?

File: docs/disabled_features.md [hunk 5]
```diff
@@ -18,6 +18,17 @@ They may be enabled by default in future versions.
 The `@` modifier lets you specify the evaluation time for instant vector selectors,
 range vector selectors, and subqueries. More details can be found [here](querying/basics.md#modifier).
 
+## Negative offset in PromQL
+
+`--enable-feature=promql-negative-offset`
+
+In contrast to the positive offset modifier, the negative offset modifier lets
+one shift a vector into the past.  An example in which one may want to use a
```
Comment (roidelapluie) [inline docs/disabled_features.md:26]:
```suggestion
one shift a vector into the future.  An example in which one may want to use a
```

File: docs/querying/basics.md [hunk 6]
```diff
@@ -204,6 +204,16 @@ The same works for range vectors. This returns the 5-minute rate that
 
     rate(http_requests_total[5m] offset 1w)
 
+For comparisons with temporal shifts forward in time, a negative offset
+can be specified:
+
+    rate(http_requests_total[5m] offset -1w)
+
+This negative offset has been disabled by default since it breaks the invariant
```
Comment (roidelapluie) [inline docs/querying/basics.md:212]:
```suggestion
This negative offset is disabled by default since it breaks the invariant
```

Comment (roidelapluie) [reply docs/querying/basics.md:212]: I think the detail about the invariant should go to the disabled feature page.

**Review by roidelapluie — APPROVED** [review]
Last nits from my side, thanks!

**Comment by roidelapluie** [comment]
Thanks!

File: docs/disabled_features.md [hunk 7]
```diff
@@ -18,6 +18,20 @@ They may be enabled by default in future versions.
...
 
+## Negative offset in PromQL
+
+This negative offset is disabled by default since it breaks the invariant
+that PromQL does not look ahead of the evaluation time for samples.
+
+`--enable-feature=promql-negative-offset`
+
+In contrast to the positive offset modifier, the negative offset modifier lets
+one shift a vector into the future.  An example in which one may want to use a
```
[the remark this answers is not in the document]
Comment (roidelapluie) [reply docs/disabled_features.md:29]:
```
For example, the following expression returns the value of
`http_requests_total` 5 minutes in the past relative to the current     
query evaluation time: 
```

We could something more similar to this wording, do you have something to propose?

[the remark this answers is not in the document]
Comment (juliusv) [reply docs/disabled_features.md:29]: That sounds confusing to me because it just sounds like you are taking a value from 5 minutes in the past (could also refer to a positive offset), and it's unclear that time-shifting happens. I would for a start just replace "future" with "past" in the line I commented on here, because that seems like a clear mistake.

[the remark this answers is not in the document]
Comment (roidelapluie) [reply docs/disabled_features.md:29]: foo offset -5m takes foo 5 minutes in the future.

[the remark this answers is not in the document]
Comment (roidelapluie) [reply docs/disabled_features.md:29]: I guess the vector is moved to the past but the vector "selector" is moved to the future.

[the remark this answers is not in the document]
Comment (juliusv) [reply docs/disabled_features.md:29]: Correct :)
